src/login.ts

```diff
@@ -6,52 +6,94 @@ import { MastoConfig } from './config';
-import { InstanceRepository } from './mastodon/v1/repositories';
+import { InstanceRepository as V1InstanceRepository } from './mastodon/v1/repositories';
+import { InstanceRepository as V2InstanceRepository } from './mastodon/v2/repositories';
 import { SerializerNativeImpl } from './serializers';
 import type { Writable } from './utils';
 import { WsNativeImpl } from './ws';
 
-export type LoginParams = {
+export interface RequestParams {
   /** URL of your instance */
   readonly url: string;
-  /** Log level to print to your console */
-  readonly logLevel?: LogType;
-  /** Access token of your account */
-  readonly accessToken?: string;
   /** Timeout in milliseconds */
   readonly timeout?: number;
   readonly defaultRequestInit?: Omit<RequestInit, 'body' | 'method'>;
-  readonly disableVersionCheck?: boolean;
   readonly disableDeprecatedWarning?: boolean;
-};
+}
 
-export const login = async (params: LoginParams): Promise<Client> => {
-  const draft: Writable<MastoConfigProps> = {
+const buildConfigProps = (params: RequestParams): MastoConfigProps => {
+  return {
     url: params.url,
     streamingApiUrl: '',
-    logLevel: params.logLevel,
-    accessToken: params.accessToken,
     timeout: params.timeout,
     defaultRequestInit: params.defaultRequestInit,
-    disableVersionCheck: params.disableVersionCheck,
     disableDeprecatedWarning: params.disableDeprecatedWarning,
   };
+};
+
+const buildRequestContext = (
+  props: MastoConfigProps,
+): {
+  serializer: SerializerNativeImpl;
+  config: MastoConfig;
+  http: HttpNativeImpl;
+} => {
   const serializer = new SerializerNativeImpl();
+  const config = new MastoConfig(props, serializer);
+  const http = new HttpNativeImpl(serializer, config);
+  return { serializer, config, http };
+};
 
-  {
-    const config = new MastoConfig(draft, serializer);
-    const http = new HttpNativeImpl(serializer, config);
-    const instance = await new InstanceRepository(http, config).fetch();
-    draft.version = new SemVer(instance.version);
-    draft.streamingApiUrl = instance.urls.streamingApi;
-  }
+export const fetchV1Instance = (
+  params: RequestParams,
+): Promise<v1.Instance> => {
+  const { http, config } = buildRequestContext(buildConfigProps(params));
+  return new V1InstanceRepository(http, config).fetch();
+};
 
-  const config = new MastoConfig(draft, serializer);
+export const fetchV2Instance = (
+  params: RequestParams,
+): Promise<v2.Instance> => {
+  const { http, config } = buildRequestContext(buildConfigProps(params));
+  return new V2InstanceRepository(http, config).fetch();
+};
+
+export interface LoginParams extends RequestParams {
+  readonly streamingApiUrl: string;
+  readonly version?: string;
+  /** Log level to print to your console */
+  readonly logLevel?: LogType;
+  /** Access token of your account */
+  readonly accessToken?: string;
+  readonly disableVersionCheck?: boolean;
+}
+
+export const createClient = (params: LoginParams): Client => {
```

Comment:
And they create a client with prev params plus `version`, `streamingApiUrl`, `accessToken`... fields.